### ISSUE TITLE
Neovim not resizing after closing iterm2 nightly split

### ISSUE THREAD

**Issue by FuzzOli87** [neovim/neovim#3378]
Hello,

When I split iterm2 while nvim is open, it resizes nvim correctly(shrinks it) however, when closing the split, the area where the split was at becomes corrupted and nvim does not resize to fill this area. This doesn't happen with regular vim nor does it happen with any other terminal app. The only way to fix it is to exit nvim entirely.

**Comment by nicolaiskogheim** [comment]
Just reinstalled nvim, and running iTerm2 nightly.

Can reproduce with my .nvimrc.

Can **not** reproduce with `nvim -u NONE`

**Comment by nicolaiskogheim** [comment]
I can't say I know what I'm doing, but here is what I have learned so far.

This is the smallest `.vimrc` that let's me reproduce the problem. Note though: I can still reproduce the bug if I remove the `vim-airline` plugin from my original `.vimrc`.
```vim
" minivimrc
call plug#begin('~/.vim/plugged')
Plug 'bling/vim-airline'                    " Statusline
call plug#end()
```

How to reproduce, method 1:
```sh
nvim -u minivimrc
<CMD-D> " Split iTerm
<CMD-D> " Split again (for a total of three windows)
<C-D> " This is in the last opened split. Or type 'exit'
<D-D> " In the second split
```
You should now see that the colors of the status line is wrong, and if not that, make a split in nvim, and see that you cannot use the whole terminal.

How to reproduce, method 2:
```sh
<CMD-D> " Split iTerm
<CMD-D> " Split again
yes " and let it run for a second
<C-C> " kill `yes`
<C-D> " to kill the split
```
This should create the same situation as above.

How NOT to reproduce:
```sh
<CMD-D> " Split iTerm
<C-D> " Kill the split
```
This alone does not seem to trigger the bug.

**Comment by FuzzOli87** [comment]
This is definitely a weird bug. I was able to record it with all of my plugins running. As you can see, when it's a clean run it doesn't refresh the view. 

However, halfway through the video, I run nvim -V12log and it updates the screen after resizing. It must be because the logger logs and then allows the next functions to continue? Not sure, but as you can see it fixes itself when it's running.

https://youtu.be/6OGfB0iaPE4

On this second video, I tested it running the .minivimrc used in the previous post. What happens here is if I close the split instantly before it loads the prompt, nvim resizes just fine. As long as it's done quickly. However if I let it set for a couple of seconds and then close it, it screws up again. 

https://www.youtube.com/watch?v=OBTzI3wAGUY

I'm running zsh as my shell by the way.

**Comment by FuzzOli87** [comment]
I'm starting to think it's an issue with the airline plugin. Here is a snippet of the v12log

```
continuing in function <SNR>10_on_window_changed..airline#update_statusline..<SNR>11_invoke_funcrefs..9
```

This function gets called when the window resizes. However, when I'm not running the log. I assume the action of writing the log gives Nvim(or airline) enough time to pick up on the change window event.

**Comment by justinmk** [comment]
The symptom, at least, is similar to https://github.com/neovim/neovim/issues/3115#issuecomment-132636353

See also https://github.com/neovim/neovim/issues/2409

**Comment by tarruda** [comment]
@FuzzOli87 Can you do a small test for me?

- go to neovim source tree(which I assume already has compiled binaries)
- type `exec build/bin/tty-test` (you will lose the current shell after it exits)

Now do the iterm split tests. The `tty-test` program is used by Neovim to test its own terminal emulator and it should print `rows y, cols: x` after the terminal window is resized.

**Comment by FuzzOli87** [comment]
@tarruda 

I couldn't find where homebrew built Neovim but I went ahead and built it from source. I ran the tty-test and it just exits the shell. Not really sure how it works after that. I opened neovim backup but I don't see any rows y, cols x at least not inside vim. However, I can see the (x,y) values in iterms' main title bar but not on a per-split basis.

It does show up though when I resize the terminal app in the main window title. Neovim still messes up but when I resize the terminal it fixes the issue. I've tried looking for an option to set it so that each split's title shows it's size but I can't find that option.

**Comment by nicolaiskogheim** [comment]
I just reproduced this in the regular Terminal.app on OSX(10.9), with the minivimrc.
NVIM 0.0.0-alpha+201509232143 (compiled Sep 25 2015 01:20:38)

It's easier to see what's happening with a window manager that resizes the window, but dragging triggers it too. Dragging also fixes it, so it can be difficult to spot.

It only happens when the window is resized in both directions. And it's only the width that is not updated, the height is somehow always right.

**Comment by sanmiguel** [comment]
I too have this problem but have found a passable workaround. It's not ideal, but... immediately before resizing, run `:messages` such that nvim still displays: `Press ENTER or type command to continue` while resizing.

I was looking into the correlation with having [vim-airline](https://github.com/bling/vim-airline) enabled and figured it might be related to the `VimResized` autocmd, so added to my .nvimrc: `au VimResized * echom 'vimresized'`. One time I accidentally left `:messages` open when resizing and the symptom went away.

I'm running nvim version:
```
NVIM 0.0.0-alpha+201510082018 (compiled Oct  9 2015 13:12:02)
Commit: 57d3a2a52fea57874d08472d0f8ee8f1bcee87c1
```
and iterm2 version `Build 2.9.20151001`

I don't know if this helps find any better detail of the problem, but since it seems to be a passable (annoying, but less so than having to restart nvim) I thought I would share.

**Comment by gonzaloserrano** [comment]
Any update on this? It happens to me several times a day, i tried with today's latest neovim HEAD and iterm2 nightly.

**Comment by omarkhan** [comment]
This is happening to me too, using 0.1.0. See screenshot below. The area on the right where the iterm2 split was displays `-- INSERT --` even when in normal mode.

<img width="640" alt="screenshot" src="https://cloud.githubusercontent.com/assets/793379/10963382/9dd4ad02-83d2-11e5-9bbd-7e247ba8728a.png">

**Comment by nhooyr** [comment]
Maybe #4154 has something to do with this? It results in half the screen going whoosh as well.

**Comment by zhimsel** [comment]
Can confirm this is happening for me as well with `nvim 0.1.2` and `iterm2 2.9.20160313`

**Comment by zhimsel** [comment]
FYI, there is a[n albiet clunky] workaround here: https://github.com/neovim/neovim/issues/3929#issuecomment-169713390

**Comment by liaoishere** [comment]
Same happen with nvim 0.1.7 and iterm2 3.0.13.
The issue goes away when `vim-airline` disabled

**Comment by justinmk** [comment]
Even if this issue were fixed, 0.1.7 will never change, so we need feedback on the development version, not 0.1.7. Can anyone here try master?

**Comment by justinmk** [comment]
Fixed by https://github.com/neovim/neovim/pull/6202
